src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.NtAuth.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Security.Authentication.ExtendedProtection;
+
+namespace System.Net.Http
+{
+    internal partial class AuthenticationHelper
+    {
+        private static Task<HttpResponseMessage> InnerSendAsync(HttpRequestMessage request, bool isProxyAuth, HttpConnection connection, CancellationToken cancellationToken)
+        {
+            return isProxyAuth ?
+                connection.SendAsyncCore(request, cancellationToken) :
+                connection.SendWithNtProxyAuthAsync(request, cancellationToken);
+        }
+
+        private static bool CheckIfProxySupportsConnectionAuth(HttpResponseMessage response)
+        {
+            if (!response.Headers.TryGetValues(KnownHeaders.ProxySupport.Descriptor, out IEnumerable<string> values))
+            {
+                return false;
+            }
+
+            foreach (string v in values)
+            {
+                if (v == "Session-Based-Authentication")
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static async Task<HttpResponseMessage> SendWithNtAuthAsync(HttpRequestMessage request, Uri authUri, ICredentials credentials, bool isProxyAuth, HttpConnection connection, CancellationToken cancellationToken)
+        {
+            HttpResponseMessage response = await InnerSendAsync(request, isProxyAuth, connection, cancellationToken).ConfigureAwait(false);
+
+            if (TryGetAuthenticationChallenge(response, isProxyAuth, authUri, credentials, out AuthenticationChallenge challenge) &&
+                (!isProxyAuth || CheckIfProxySupportsConnectionAuth(response)))
+            {
+                if (challenge.AuthenticationType == AuthenticationType.Negotiate || 
+                    challenge.AuthenticationType == AuthenticationType.Ntlm)
+                {
+                    string challengeData = challenge.ChallengeData;
+
+                    string spn = "HTTP/" + authUri.IdnHost;
```

Comment:
Are SPNs supposed to be punycode or unicode?

Reply:
My understanding is it should match the hostname we're connecting to, so punycode.

Reply:
Please test it. SPNs and host names are only vaguely related.